lib/maths/CBoostedTreeImpl.cc

```diff
@@ -808,16 +817,18 @@ CBoostedTreeImpl::trainTree(core::CDataFrame& frame,
     workspace.reinitialize(m_NumberThreads, candidateSplits, m_Loss->numberParameters());
 
     TNodeVec tree(1);
-    tree.reserve(2 * maximumTreeSize + 1);
+    // since number of leaves in a perfect binary tree is (numberInternalNodes+1)
+    // the total number of nodes in a tree is (2*numberInternalNodes+1)
+    tree.reserve(2 * maximumNumberInternalNodes + 1);
 
     // Sampling transforms the probabilities. We use a placeholder outside
     // the loop adding nodes so we only allocate the vector once.
     TDoubleVec featureSampleProbabilities{m_FeatureSampleProbabilities};
     TSizeVec featureBag;
     this->featureBag(featureSampleProbabilities, featureBag);
 
-    TLeafNodeStatisticsPtrQueue leaves(maximumTreeSize / 2 + 3);
-    leaves.push_back(std::make_shared<CBoostedTreeLeafNodeStatistics>(
+    TLeafNodeStatisticsPtrQueue splitCandidateTreeNodes(maximumNumberInternalNodes / 2 + 3);
```

Comment:
This is the current working set of tree leaves' statistics, i.e. notice we always pop the node after splitting it, alternatively we only ever consider splitting a leaf of the current tree. Therefore, I wouldn't drop this from the name. Since you know have a constraint such that it isn't all leaves how about calling it `splittableLeaves`?
```suggestion
    TLeafNodeStatisticsPtrQueue splittableLeaves(maximumNumberInternalNodes / 2 + 3);
```

Reply:
Renamed